src/timer.rs

```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
-                    self.set_ctl(0x01, 0x02);
+                    self.tim.tax_ctl.modify(|_, w| w.taclr().set_bit());
                     self.count = 0;
                 }
 
                 #[inline]
                 fn setup_timer(&self, count: Count) -> bool {
+                    use TimerUnit::*;
+
                     match count.1 {
-                        TimerUnit::Hertz => self.setup_frequency(count.0),
-                        TimerUnit::Kilohertz => self.setup_frequency(1000*count.0),
-                        TimerUnit::Milliseconds => self.setup_period(count.0),
-                        TimerUnit::Seconds => self.setup_period(1000*count.0),
+                        Hertz | Milliseconds => self.setup_count(count),
+                        Kilohertz => self.setup_count(Count(count.0 * 1000, Hertz)),
+                        Seconds => self.setup_count(Count(count.0 * 1000, Milliseconds)),
                     }
                 }
 
-                fn set_clock_source(&self, value: u8, prescaler: u8) {
-                    self.clear_ctl(0x0F, 0x06);
-                    self.set_ctl((value | prescaler) as u16, 0x06);
-                }
+                #[inline]
+                fn setup_count(&self, count: Count) -> bool {
+                    use TimerUnit::*;
+                    use ClockSource::*;
 
-                fn setup_period(&self, period: u32) -> bool {
-                    let smclk_period = period*(self.clocks.smclk.0/1000);
-                    let aclk_period = (period*self.clocks.aclk.0)/1000;
                     let max_period = MAX_PRESCALER * MAX_COUNT;
-
-                    if smclk_period < max_period {
-                        let prescaler = self.get_prescaler((smclk_period / MAX_COUNT) as u8 + 1);
-                        let count = ((period * self.clocks.smclk.0) / (prescaler[2] as u32 * 1000)) as u16;
-                        self.set_clock_source(0x08, prescaler[1]);
-                        self.setup_count(count);
-                        true
-                    } else if aclk_period < max_period {
-                        let prescaler = self.get_prescaler((aclk_period / MAX_COUNT) as u8 + 1);
-                        let count = ((period * self.clocks.aclk.0) / (prescaler[2] as u32 * 1000)) as u16;
-                        self.set_clock_source(0x04, prescaler[1]);
-                        self.setup_count(count);
-                        true
+                    let aclk_ratio: u32;
+                    let real_prescaler : ClockSourcePrescaler;
+                    let smclk_ratio : u32;
+                    let tick_count : u16;
+                    let mut count_ratio : u32;
+                    let mut clock_source : ClockSource;
+                    let mut is_config: bool;
+                    let mut min_prescaler : u8;
+
+                    is_config = false;
+                    count_ratio = 0;
+                    clock_source = Smclk;
```

Comment:
What does `is_config` means in this fn context?

```suggestion
                    let aclk_ratio: u32;
                    let smclk_ratio : u32;
                    let mut count_ratio : u32  = 0;
                    let mut clock_source : ClockSource = Smclk;
                    let mut is_config: bool = false;
```

Reply:
The try_start function return an error (Error::Unreachable) if the counting value is out of the bounds (is_config == false). Return true or false will take the same effect.

Reply:
I think it's simpler to just return true at the end of the block and kill this val